### ISSUE TITLE
Discussion of explicit typeclass dictionaries

### ISSUE THREAD

**Issue by tomjaguarpaw** [tomjaguarpaw/ad#1]
@Lysxia @tysonzero I have implemented a PoC for the explicit typeclass dictionaries stuff.  I was going to think it through some more before bringing it to your attention but I never got round to it so you may as well see it in its basic state.

I will improve it some more in the future.  Feel free to discuss on this ticket or submit PRs if you have any good ideas.  You can find it here: https://github.com/tomjaguarpaw/ad/blob/master/Class/Class.hs

(This repository is called `ad` because it used to be for automatic differentiation experimentation.  It now stands for "assorted delights".)

**Comment by Lysxia** [comment]
Thanks for this! I'm happy to see the idea move forward.

One thought is that `Class` could contain the whole instance head in one argument:

```
class Superclass a => Class a where
  methods :: a

instance Class (Semigroup [a]) where
  methods = listSemigroup
```

This generalizes nicely to multiparam classes. And for whatever the following remark is worth, that's `Default`!

**Comment by tomjaguarpaw** [comment]
Yes, that's the `Sat` approach mentioned here: http://h2.jaguarpaw.co.uk/posts/scrap-all-your-typeclasses-but-one/

I'm not really sure what the consequences of that approach are.  Perhaps one of us could try rewriting `Class.hs` using that approach and see if anything breaks.

**Comment by tysonzero** [comment]
Awesome! I have a couple thoughts.

One thing is that I think you probably are going to want `MonoidD` to contain `<>` as well. I know it does kind of mean multiple sources of truth, but at the same time `MonoidD` as it stands is relatively useless as a standalone type.

Lets say we want to just define a `BooleanRing` type, and derive an entire ring and ordering hierarchy off of casting / converting this type to `Monoid` and `Lattice` and so on. We can only do this if `BooleanRing` actually contains all the functionality inherent to a `BooleanRing`.

Similarly `MonoidD` as it stands does not give us `foldX :: FoldableD f -> MonoidD a -> f a -> a`, we'd need `foldX :: FoldableD f -> MonoidD a -> SemigroupD -> f a -> a` which also does not really hint that `MonoidD` and `SemigroupD` need to be compatible.

As for the `Default` thing. I have thought about that before and I don't think it quite works as a fundamental building block. Oftentimes types will have multiple good ways to implement a certain dictionary, such as `0, +` and `1, *` for `Monoid Int`. So having `AddMonoid`, `MultMonoid` and the default `Monoid` all share the same underlying structure is nice but incompatible with only using a single class. It also doesn't seem to work with TypeFamilies, so useful classes like `MonoFoldable` and `Ixed` and so on would have worse type inference than they do now.

With that said it could be useful as another piece of all of this, such as `class Default (MonoidD a) => Monoid a` or even `type Monoid a = Default (MonoidD a)`. If it weren't for the type inference aspects I could even see us avoiding the awkward `D` suffix by just having people use `Default (Monoid a)` and not having an explicit `Monoid` class except for explicit variants like `AddMonoid`.

It would be nice to try and find a way to avoid the `D` suffix as it makes the underlying algebraic structures feel like second class citizens. For classes with an `Is` prefix or `able` suffix things are fine as we can just have `Fold` be the type and `Foldable` be the class, but for others we need to put in more effort, perhaps `IsMonoid` or `HasMonoid` for the class name and `Monoid` for the type. `Monoid` fundamentally should be the type and not the class from a mathematical perspective at least.

Another thing I want to think about is how we want to handle laws. Classes and instances kind of get around this issue by having it just be a community expectation to watch out for and obey laws when defining typeclass instances. People are far more willing to just throw whatever typechecks into functions and constructors though, and we really want to make it hard to accidentally make a `MonoidD` that doesn't obey the laws. Of course this is Haskell and not Agda, so I'm not really asking for proofs in the traditional sense, but some sort of syntactically obvious assertion that perhaps comes with easy optional quickCheck-ing or optional proving (when we have dependent haskell anyway) could be cool. So you would end with something like `intAddMonoid = mkMonoid { (<>) = (+), mempty = 0, isAssociative = Assert, isIdentity = Assert}` or similar.

**Comment by Lysxia** [comment]
At a less technical level, I see at least three ways to think about/explore the proposal, not necessarily disjoint:

1. different usage patterns for existing language feature (that seems to be the only way for any early PoC such as this, where code can be considered "working")

2. a brand new language feature (what if we could redo type classes)

3. a new language feature (maybe many), but with a story for compatibility with the existing ecosystem (even if only to admit that some things have to break)

On various details (to be) discussed here I would have very different answers depending on which of these directions I'm facing. I don't really know whether that's the right way to think about our objectives, I may just be rambling, but that seemed useful to mention to help clarify what we want to do, where we want to go. I guess we'll figure out part of it along the way. Feel free to tell if you have very different point of views about how to steer the boat.

**Comment by tysonzero** [comment]
So I guess I'm mostly thinking (1).

Although my desire for extensible rows/records/variants often causes me to leak a bit outside of the existing language, similarly I did mention dependent typing briefly in my post above.

In my original simplifying typeclasses write-up the only feature I propose is lightweight syntax sugar that we can just ignore for now.

I think we can do quite a lot within the current features of the language by just sticking to single-member typeclasses and going from there, and I think that would have the most practical value, at least in the near future.

**Comment by tomjaguarpaw** [comment]
The `Def`approach seems to go through very nicely! https://github.com/tomjaguarpaw/ad/pull/2

For some reason I'm surprised by this.

**Comment by tomjaguarpaw** [comment]
For reference, here's @tysonzero's related GHC proposal: https://github.com/ghc-proposals/ghc-proposals/pull/324 (Allow direct access to underlying concrete class dictionaries).
